subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/DefaultProjectSchemaProvider.kt

```diff
@@ -149,8 +150,17 @@ val KClass<*>.firstKotlinPublicOrSelf
 
 private
 val KClass<*>.firstKotlinPublicOrNull: KClass<*>?
-    get() = takeIf { visibility == KVisibility.PUBLIC }
-        ?: superclasses.firstNotNullResult { it.firstKotlinPublicOrNull }
+    get() = takeIf { isKotlinVisible && visibility == KVisibility.PUBLIC }
+        ?: java.interfaces.firstNotNullResult { it.kotlin.firstKotlinPublicOrNull }
+        ?: (java.superclass as? Class<*>)?.kotlin?.firstKotlinPublicOrNull
+
+
+private
+val KClass<*>.isKotlinVisible: Boolean
+    get() = Modifier.isPublic(java.modifiers) &&
+        !java.isLocalClass &&
+        !java.isAnonymousClass &&
```

Comment:
⭕️ I wonder if it would be possible to recover the public SAM type from a lambda class.

Reply:
See https://github.com/gradle/gradle/issues/10766